### PR TITLE
X86 fastcall convention support

### PR DESCRIPTION
**Pull request by 0dvictor**
Introducing a new linkage TR_FastCall. It is mapped to the following calling convention in each platform:
64-bit Linux: System V AMD64 ABI (same as TR_System)
64-bit Windows: Microsoft x64 calling convention (same as TR_System)
32-bit Linux: fastcall with sseregparm
32-bit Windows: fastcall

Signed-off-by: Victor Ding <dvictor@ca.ibm.com>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
FYI...  @mgaudet is working on extern call support with a user-specified linkage in Tril (#1910) that should make creating unit test cases to exercise this kind of code relatively straightforward.  I would like to see some unit test cases written with it once its ready.

**Comment by mgaudet** [comment]
So, given the linkage convention enum is shared across architectures, is there a particular expectation for `TR_Fastcall` linkage? Or is just a linkage expected to be 'faster' than the default linkage? Would aliasing this to `TR_SystemLinkage` break expectations? 

A naming aside: Not being a linkage expert, squinting, it seems like this code is more trying to implement something similar to the [Microsoft vectorcall linkage](https://en.wikipedia.org/wiki/X86_calling_conventions#Microsoft_vectorcall), rather than fast call.

File: compiler/x/codegen/RealRegisterManager.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2017 IBM Corp. and others
```
Comment (fjeremic) [inline compiler/x/codegen/RealRegisterManager.hpp]: This looks like a new file. The initial copyright date seems incorrect. Similarly elsewhere in this PR.

File: compiler/x/codegen/RealRegisterManager.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      for (uint8_t i = (uint8_t)TR::RealRegister::NoReg; i < (uint8_t)TR::RealRegister::NumRegisters; i++)
+         {
+         if (_Registers[i] != NULL)
+            {
+            _cg->stopUsingRegister(_Registers[i]);
+            }
+         }
+      }
+   // Find or allocate a virtual register which is bind to the given real register
+   TR::Register* Use(TR::RealRegister::RegNum RealRegister)
```
Comment (fjeremic) [inline compiler/x/codegen/RealRegisterManager.hpp]: `Use` seems to be orthogonal to the camelCase naming convention used throughout the entire compiler. Similarly for `BuildRegisterDependencyConditions`, `NumberOfRegistersInUse`, etc. We should try to stay consistent if possible.

**Comment by 0dvictor** [comment]
The implementation is actually gcc's `fastcall` with `sseregparm`, which is slightly different from Microsoft's `vectorcall`. `vectorcall` is only supported by MSVS 2013 or newer; however OMR is required to support MSVS 2010.
Linkages on X86-32 is such a complicated topic. There are too many choices and none of them is the standard. For OMR, `TR_System` is mapped to `cdecl`, which is used by Linux APIs and Microsoft's MSVCRT. `cdecl` requires floating point return value being passed by X87 register ST(0) even if both caller and callee use only SSE/AVX. Therefore I'm proposing OMR to adopt `fastcall` with `sseregparam`, to avoid the cost of unnecessary value movements between SSE/AVX and X87, for OMR's internal calls and helper calls.
On X86-64, `TR_FastCall` is identical to `TR_System`, and I'm expecting they are identical on ARM, Power, and Z.

File: compiler/x/codegen/RealRegisterManager.hpp (hunk 1 above)
Comment (0dvictor) [reply compiler/x/codegen/RealRegisterManager.hpp]: Thanks for pointing it out, it has been updated.

File: compiler/x/codegen/RealRegisterManager.hpp (hunk 2 above)
Comment (0dvictor) [reply compiler/x/codegen/RealRegisterManager.hpp]: Sure, I've updated them.

File: compiler/x/amd64/codegen/AMD64FastCallConfig.hpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2017 IBM Corp. and others
```
Comment (mstoodle) [inline compiler/x/amd64/codegen/AMD64FastCallConfig.hpp]: Are these files marked as copyright 2000 because they contain excerpts from other files that have copyright 2000?

**Comment by mstoodle** [comment]
@0xdaryl @0dvictor #1910 was merged a while back. Is the expectation now that tril unit tests should be written to test this fastcall linkage before this PR is merged?

File: compiler/x/amd64/codegen/AMD64FastCallConfig.hpp (hunk 3 above)
Comment (0dvictor) [reply compiler/x/amd64/codegen/AMD64FastCallConfig.hpp]: Thank you for point it out. In this case, it was a mistake; X86FastCallLinkage.hpp and X86FastCallLinkage.cpp contain modified code from OpenJ9, which have copyright 2000.

Comment (0dvictor) [reply compiler/x/amd64/codegen/AMD64FastCallConfig.hpp]: Updated.

**Comment by mstoodle** [comment]
This PR blocks another PR at OpenJ9 and they've both been open for several months without much recent documented progress. Would you please outline the plan here @0dvictor @andrewcraik ?

**Comment by 0dvictor** [comment]
The functional change is ready; however, its related tests are not there yet. I don't know whether we should wait for the tests or get this in first.

**Comment by mstoodle** [comment]
So, I guess one on-point question: can you explain what you've done to verify that the functional implementation is correct?

**Comment by 0dvictor** [comment]
I have test the functionalities with https://github.com/eclipse/openj9/pull/477 . The combination passed all IBM's internal Java tests.

**Comment by 0dvictor** [comment]
I've created some sanity tests using ones for system linkage as template. Note that tril based OMR tests do not cover all code paths of fastcall linkage. It may or may not be a limitation that tril based OMR tests always generate a caller-callee pair such that their addresses are more than 2G apart. Therefore, these tests only exercise the path that a trampoline is used. Fortunately, OpenJ9 always exercises the other path that caller and callee are less than 2G apart. I hope by tests on both OMR and OpenJ9 side, it can be confidence that this PR is correct and good to go.

File: compiler/x/codegen/X86FastCallLinkage.cpp [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+   // Build parameters, parameters in _Params are Right-to-Left (RTL)
+   TR::Register* IntParams[NumberOfIntParamRegisters] = {};
+   TR::Register* FloatParams[NumberOfFloatParamRegisters] = {};
+   TR_Stack<TR::Register*> StackParams(cg()->trMemory());
+
+   size_t IntParamIndex = 0;
+#ifndef WINDOWS
+   size_t FloatParamIndex = 0;
+#else
+   auto& FloatParamIndex = IntParamIndex;
```
Comment (mstoodle) [inline compiler/x/codegen/X86FastCallLinkage.cpp]: I'm not a big fan of this construct. While it moves the #ifdef out of the loop, it makes that loop code somewhat harder to understand (the parameter indices could be the same variable or not, depending on the platform). This is an example where I think I'd actually prefer to see the #ifdef code inside the loop (gasp! I know...)

File: compiler/x/codegen/X86FastCallLinkage.hpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
```
Comment (mstoodle) [inline compiler/x/codegen/X86FastCallLinkage.hpp]: another out of date SPDX identifier.

File: compiler/x/codegen/X86FastCallLinkage.hpp [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+#include "codegen/Linkage.hpp"
+#include "env/jittypes.h"
+
+namespace TR { class CodeGenerator; }
+namespace TR { class Instruction; }
+namespace TR { class LabelSymbol; }
+namespace TR { class Node; }
+namespace TR { class RegisterDependencyConditions; }
+
+#define IMCOMPLETELINKAGE  "This class is only used to generate call-out sequence but no call-in sequence, so it is not used as a complete linkage."
```
Comment (mstoodle) [inline compiler/x/codegen/X86FastCallLinkage.hpp]: typo: INCOMPLETELINKAGE (note second letter)

File: compiler/x/i386/codegen/IA32FastCallConfig.hpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
```
Comment (mstoodle) [inline compiler/x/i386/codegen/IA32FastCallConfig.hpp]: yup, another SPDX id needs to be updated :)

**Comment by mstoodle** [comment]
@0dvictor what do you think about changing that remaining loop structure?

**Comment by mstoodle** [comment]
I can see what you're saying. I'm just not sure (as a "reader" of that code) that I agree. 
But I'm not sure I feel strongly enough about it to require you to change it. At the very least, an informative comment is warranted.

I'll defer to a second opinion from @0xdaryl on whether he thinks this variable aliasing for the Windows platform is better or worse than putting the #ifdef inside the loop (or maybe he'll suggest a third option for even more fun).

**Comment by mstoodle** [comment]
Ping to @0xdaryl for his opinion :)

**Comment by andrewcraik** [comment]
@0dvictor, @0xdaryl and @mstoodle this looks to have gone stale - do we know what the next steps are?

**Comment by mstoodle** [comment]
Still marked as WIP @0dvictor what do you think of closing this PR and then we can look at it with fresh eyes once ready for merging (though I'd still like to see a review from @0xdaryl who's currently on vacation).

**Comment by mstoodle** [comment]
Thanks @0dvictor please @ ping me when the PR is ready.

**Comment by 0dvictor** [comment]
@mstoodle FYI, this changeset has been rebased and re-verified.

File: compiler/x/codegen/X86FastCallLinkage.cpp (hunk 4 above)
Comment (mstoodle) [reply compiler/x/codegen/X86FastCallLinkage.cpp]: I am still not a huge fan, but I can't argue that it doesn't match how the Windows ABI maps registers. I'll withdraw the opinion as not important enough to block the change.

**Comment by mstoodle** [comment]
genie-omr build all

**Comment by 0dvictor** [comment]
For sure, I'll update the commit message to the following once tests pass:

Introducing a new linkage TR_FastCall. It is mapped to the following calling convention in each platform:
64-bit Linux: System V AMD64 ABI (same as TR_System)
64-bit Windows: Microsoft x64 calling convention (same as TR_System)
32-bit Linux: fastcall with sseregparm
32-bit Windows: fastcall

**Comment by mstoodle** [comment]
@0dvictor looks like a possibly valid failure ...

**Comment by 0dvictor** [comment]
The failure occurred at the trampoline to call an address that not fit into a 32-bit RIP relative. It is never used in downstream projects that I run all tests, and did not occur in my local OMR sanity test. It has been fixed.

**Comment by 0dvictor** [comment]
@mstoodle do you mind take a look at the updated change?

**Comment by mstoodle** [comment]
genie-omr build all

**Comment by mstoodle** [comment]
I'm ok with the commits as they are, but I'm not the x86 expert. @0xdaryl would you please review the x86 details of this change?

**Comment by 0xdaryl** [comment]
I am aiming to give this a comprehensive review by EOW.

**Comment by mstoodle** [comment]
Still pondering @0xdaryl :) ?

**Comment by Leonardo2718** [comment]
ping @0xdaryl

**Comment by Leonardo2718** [comment]
@0dvictor there seem to be a few merge conflicts that need to be resolved.

**Comment by 0dvictor** [comment]
The conflicts have been resolved.

**Comment by Leonardo2718** [comment]
Seems like `getX86RealRegister` is being used instead of `getRealRegister`. An artefact of merge conflict resolution?

**Comment by 0dvictor** [comment]
Good catch, fixed.

**Comment by Leonardo2718** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
My principal concerns with this PR are that it's being done in a different way than how the other linkages are implemented and introduces a subtly different coding style than found in other parts of the x86 code generator (or the compiler for that matter).  For instance, this implementation does not use the LinkageProperties structure to describe the registers.  While I do think the LinkageProperties needs an overhaul (it's about to celebrate its 20th birthday!), I'm not convinced that the solution proposed in this PR is entirely the way to go (I do like some of the ideas though), and I am hesitant to introduce another design variant into the codebase without a plan to bring all linkages to use a common design (not just x86, but across all platforms).  I will ultimately be creating an issue to revisit the LinkageProperties design as part of #2082.

While my concerns could be addressed in this PR, given that the author is no longer actively involved in the project and is unlikely to make progress with moving this PR along, I am closing it.
